### PR TITLE
Introduce ConceptID object to tighten up code base

### PR DESCRIPTION
**Pull request by fppt**
- Introduces new ConceptID object which is used instead of String
- Propagate the changes upwards.

### CODE REVIEW

File: grakn-core/src/main/java/ai/grakn/concept/ConceptId.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * <p>
+ *     A class which represents an id of any {@link Concept} in the {@link ai.grakn.GraknGraph}.
+ *     Also contains a static method for producing concept IDs from Strings.
+ * </p>
+ *
+ * @author fppt
+ */
+public class ConceptId implements Comparable<ConceptId> {
+    private static Map<String, ConceptId> conceptIds = new HashMap();
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/ConceptId.java]: This might be a good idea, but it could get very big over time.

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java [hunk 2]
```diff
@@ -72,11 +73,11 @@ public String newState(String taskName, String createdBy, Date runAt, Boolean re
 
         Optional<String> result = attemptCommitToSystemGraph((graph) -> {
             InsertQuery query = graph.graql().insert(state);
-            String id = query.stream().findFirst().get().get(TASK_VAR).getId();
+            ConceptId id = query.stream().findFirst().get().get(TASK_VAR).getId();
 
             LOG.debug("Created " + graph.getConcept(id));
 
-            return id;
+            return id.getValue();
```
Comment (aelred) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java]: Do we have to extract the value here, instead of keeping it as a `ConceptId`?

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java [hunk 3]
```diff
@@ -213,7 +214,7 @@ private TaskState instanceToState(GraknGraph graph, Instance instance){
             for (Map<String, Concept> m : res) {
                 Concept c = m.values().stream().findFirst().orElse(null);
                 if (c != null) {
-                    String id = c.getId();
+                    String id = c.getId().getValue();
```
Comment (aelred) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java]: Same as above here

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/IdFragment.java [hunk 4]
```diff
@@ -44,7 +45,7 @@ public void applyTraversal(GraphTraversal<Vertex, Vertex> traversal) {
 
     @Override
     public String getName() {
-        return "[id:" + idToString(id) + "]";
+        return "[id:" + idToString(id.getValue()) + "]";
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/IdFragment.java]: Maybe we can make the `idToString` method accept a `ConceptId`

Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/IdFragment.java]: In fact, we might be able to make this the `ConceptId#toString` method (or some other method on `ConceptId`)

File: grakn-graql/src/main/java/ai/grakn/graql/internal/parser/QueryVisitor.java [hunk 5]
```diff
@@ -466,7 +467,7 @@ public Var visitVarPattern(GraqlParser.VarPatternContext ctx) {
 
     @Override
     public UnaryOperator<Var> visitPropId(GraqlParser.PropIdContext ctx) {
-        return var -> var.id(visitId(ctx.id()));
+        return var -> var.id(ConceptId.of(visitId(ctx.id())));
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/parser/QueryVisitor.java]: We can make `visitId` return a `ConceptId` instead of a `String`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java [hunk 6]
```diff
@@ -94,7 +95,7 @@
         varsById = ImmutableMap.copyOf(
                 vars.stream()
                         .filter(var -> var.getId().isPresent())
-                        .collect(Collectors.groupingBy(var -> var.getId().get()))
+                        .collect(Collectors.groupingBy(var -> var.getId().get().getValue()))
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java]: We can definitely make `varsById` be of type `ImmutableMap<ConceptId, List<VarAdmin>>` now.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/Binary.java [hunk 7]
```diff
@@ -18,18 +19,18 @@
  *
  */
 public abstract class Binary extends BinaryBase {
-    private Predicate predicate = null;
+    private IdPredicate predicate = null;
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/Binary.java]: What's the reason for this change?

Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/Binary.java]: It was always an Id predicate being injected into a `Binary` Class, so I assumed the change would be safe and I needed the `IdPredicate` so I could get access to the ConceptID without changing the Base Predicate class.

Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/Binary.java]: it actually makes sense cause Binaries always have IdPredicates specifically

Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/Binary.java]: Woohoo I did a good change in reasoner!! Now on to my bad changes :)

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java (hunk 6 above)
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java]: Just did it actually. Ended up solving a few failing tests. Yay!

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java (hunk 2 above)
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java]: The class `GraknStateStorage` extends `StateStorage` which has two implementations. the one based on using a grakn graph, and the other based on using an in memory map which uses only strings. This also ties in with a Zookeeper class which only uses Strings.

Basically putting everything on `ConceptId`feels wrong since only one implementation uses a Grakn Graph. There is definitely a way to improve it though. Generics on the interface is the way to go I think.

File: grakn-core/src/main/java/ai/grakn/concept/ConceptId.java (hunk 1 above)
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/concept/ConceptId.java]: Kill it for now then? I was not sure of it myself anyway.

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java (hunk 2 above)
Comment (aelred) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstorage/GraknStateStorage.java]: @fppt and I have agreed to do this with a generic in a future PR.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/parser/QueryVisitor.java [hunk 8]
```diff
@@ -333,7 +334,8 @@ public PathQuery visitPath(GraqlParser.PathContext ctx) {
 
         if (ctx.PERSIST() != null) {
             if (ctx.id() != null) {
-                cluster = cluster.persist(visitId(ctx.id()));
+                //TODO: Possible a bug that this is asking for the Type name but we extracting the ID. Temporary workaround using getValue()
+                cluster = cluster.persist(visitId(ctx.id()).getValue());
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/parser/QueryVisitor.java]: Paging @duckofyork so he's aware of this!

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/PathQueryImpl.java [hunk 9]
```diff
@@ -62,7 +63,8 @@
         ComputerResult result;
         try {
             result = getGraphComputer().compute(
-                    new ShortestPathVertexProgram(subTypeNames, sourceId, destinationId),
+                    //TODO: Look into passing sourceId and destinationId as ConceptId. Not possible right now because it's not serializable
+                    new ShortestPathVertexProgram(subTypeNames, sourceId.getValue(), destinationId.getValue()),
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/PathQueryImpl.java]: My understanding of serialisable isn't very good, but I think this is as easy as just `ConceptId implements Serializable`.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/Atom.java [hunk 10]
```diff
@@ -48,14 +49,14 @@
...
-    protected String typeId = null;
+    protected ConceptId typeId = null;
 
     protected Atom(VarAdmin pattern) { this(pattern, null);}
     protected Atom(VarAdmin pattern, Query par) { super(pattern, par);}
     protected Atom(Atom a) {
         super(a);
         this.type = a.type;
-        this.typeId = a.typeId;
+        this.typeId = a.getTypeId() != null? ConceptId.of(a.getTypeId().getValue()) : null;
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/Atom.java]: You can replace `ConceptId.of(a.getTypeId().getValue()` with `a.getTypeId()`.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/MultiPredicateBinary.java [hunk 11]
```diff
@@ -53,11 +54,11 @@ protected MultiPredicateBinary(VarAdmin pattern, Set<Predicate> preds, Query par
 
     protected MultiPredicateBinary(MultiPredicateBinary a) {
         super(a);
-        a.getMultiPredicate().forEach(multiPredicate::add);
-        this.typeId = extractTypeId(atomPattern.asVar());
+        a.getMultiPredicate().forEach(pred -> multiPredicate.add((Predicate) AtomicFactory.create(pred, getParentQuery())));
+        this.typeId = a.getTypeId() != null? ConceptId.of(a.getTypeId().getValue()) : null;
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/MultiPredicateBinary.java]: You can replace `ConceptId.of(a.getTypeId().getValue()` with `a.getTypeId()`.
